### ISSUE TITLE
Gradients do not exist for variables

### ISSUE THREAD

**Issue by fakerbrother** [kpe/bert-for-tf2#46]
hello everyone, when I start training my custom model, I meet two problem, [this](https://github.com/kpe/bert-for-tf2/issues/44), another one is:

`W0120 15:20:05.422765 12180 optimizer_v2.py:1029] Gradients do not exist for variables ['bert/embeddings/word_embeddings/embeddings:0', 'bert/embeddings/position_embeddings/embeddings:0', 'bert/embeddings/LayerNorm/gamma:0', 'bert/embeddings/LayerNorm/beta:0', 'bert/encoder/layer_0/attention/self/query/kernel:0', 'bert/encoder/layer_0/attention/self/query/bias:0', 'bert/encoder/layer_0/attention/self/key/kernel:0', 'bert/encoder/layer_0/attention/self/key/bias:0', 'bert/encoder/layer_0/attention/self/value/kernel:0', 'bert/encoder/layer_0/attention/self/value/bias:0', 'bert/encoder/layer_0/attention/output/dense/kernel:0', 'bert/encoder/layer_0/attention/output/dense/bias:0', 'bert/encoder/layer_0/attention/output/LayerNorm/gamma:0', 'bert/encoder/layer_0/attention/output/LayerNorm/beta:0', 'bert/encoder/layer_0/intermediate/kernel:0', 'bert/encoder/layer_0/intermediate/bias:0', 'bert/encoder/layer_0/output/dense/kernel:0', 'bert/encoder/layer_0/output/dense/bias:0', 'bert/encoder/layer_0/output/LayerNorm/gamma:0', 'bert/encoder/layer_0/output/LayerNorm/beta:0', 'bert/encoder/layer_1/attention/self/query/kernel:0', 'bert/encoder/layer_1/attention/self/query/bias:0', 'bert/encoder/layer_1/attention/self/key/kernel:0', 'bert/encoder/layer_1/attention/self/key/bias:0', 'bert/encoder/layer_1/attention/self/value/kernel:0', 'bert/encoder/layer_1/attention/self/value/bias:0', 'bert/encoder/layer_1/attention/output/dense/kernel:0', 'bert/encoder/layer_1/attention/output/dense/bias:0', 'bert/encoder/layer_1/attention/output/LayerNorm/gamma:0', 'bert/encoder/layer_1/attention/output/LayerNorm/beta:0', 'bert/encoder/layer_1/intermediate/kernel:0', 'bert/encoder/layer_1/intermediate/bias:0', 'bert/encoder/layer_1/output/dense/kernel:0', 'bert/encoder/layer_1/output/dense/bias:0', 'bert/encoder/layer_1/output/LayerNorm/gamma:0', 'bert/encoder/layer_1/output/LayerNorm/beta:0', 'bert/encoder/layer_2/attention/self/query/kernel:0', 'bert/encoder/layer_2/attention/self/query/bias:0', 'bert/encoder/layer_2/attention/self/key/kernel:0', 'bert/encoder/layer_2/attention/self/key/bias:0', 'bert/encoder/layer_2/attention/self/value/kernel:0', 'bert/encoder/layer_2/attention/self/value/bias:0', 'bert/encoder/layer_2/attention/output/dense/kernel:0', 'bert/encoder/layer_2/attention/output/dense/bias:0', 'bert/encoder/layer_2/attention/output/LayerNorm/gamma:0', 'bert/encoder/layer_2/attention/output/LayerNorm/beta:0', 'bert/encoder/layer_2/intermediate/kernel:0', 'bert/encoder/layer_2/intermediate/bias:0', 'bert/encoder/layer_2/output/dense/kernel:0', 'bert/encoder/layer_2/output/dense/bias:0', 'bert/encoder/layer_2/output/LayerNorm/gamma:0', 'bert/encoder/layer_2/output/LayerNorm/beta:0', 'bert/encoder/layer_3/attention/self/query/kernel:0', 'bert/encoder/layer_3/attention/self/query/bias:0', 'bert/encoder/layer_3/attention/self/key/kernel:0', 'bert/encoder/layer_3/attention/self/key/bias:0', 'bert/encoder/layer_3/attention/self/value/kernel:0', 'bert/encoder/layer_3/attention/self/value/bias:0', 'bert/encoder/layer_3/attention/output/dense/kernel:0', 'bert/encoder/layer_3/attention/output/dense/bias:0', 'bert/encoder/layer_3/attention/output/LayerNorm/gamma:0', 'bert/encoder/layer_3/attention/output/LayerNorm/beta:0', 'bert/encoder/layer_3/intermediate/kernel:0', 'bert/encoder/layer_3/intermediate/bias:0', 'bert/encoder/layer_3/output/dense/kernel:0', 'bert/encoder/layer_3/output/dense/bias:0', 'bert/encoder/layer_3/output/LayerNorm/gamma:0', 'bert/encoder/layer_3/output/LayerNorm/beta:0', 'bert/encoder/layer_4/attention/self/query/kernel:0', 'bert/encoder/layer_4/attention/self/query/bias:0', 'bert/encoder/layer_4/attention/self/key/kernel:0', 'bert/encoder/layer_4/attention/self/key/bias:0', 'bert/encoder/layer_4/attention/self/value/kernel:0', 'bert/encoder/layer_4/attention/self/value/bias:0', 'bert/encoder/layer_4/attention/output/dense/kernel:0', 'bert/encoder/layer_4/attention/output/dense/bias:0', 'bert/encoder/layer_4/attention/output/LayerNorm/gamma:0', 'bert/encoder/layer_4/attention/output/LayerNorm/beta:0', 'bert/encoder/layer_4/intermediate/kernel:0', 'bert/encoder/layer_4/intermediate/bias:0', 'bert/encoder/layer_4/output/dense/kernel:0', 'bert/encoder/layer_4/output/dense/bias:0', 'bert/encoder/layer_4/output/LayerNorm/gamma:0', 'bert/encoder/layer_4/output/LayerNorm/beta:0', 'bert/encoder/layer_5/attention/self/query/kernel:0', 'bert/encoder/layer_5/attention/self/query/bias:0', 'bert/encoder/layer_5/attention/self/key/kernel:0', 'bert/encoder/layer_5/attention/self/key/bias:0', 'bert/encoder/layer_5/attention/self/value/kernel:0', 'bert/encoder/layer_5/attention/self/value/bias:0', 'bert/encoder/layer_5/attention/output/dense/kernel:0', 'bert/encoder/layer_5/attention/output/dense/bias:0', 'bert/encoder/layer_5/attention/output/LayerNorm/gamma:0', 'bert/encoder/layer_5/attention/output/LayerNorm/beta:0', 'bert/encoder/layer_5/intermediate/kernel:0', 'bert/encoder/layer_5/intermediate/bias:0', 'bert/encoder/layer_5/output/dense/kernel:0', 'bert/encoder/layer_5/output/dense/bias:0', 'bert/encoder/layer_5/output/LayerNorm/gamma:0', 'bert/encoder/layer_5/output/LayerNorm/beta:0', 'bert/encoder/layer_6/attention/self/query/kernel:0', 'bert/encoder/layer_6/attention/self/query/bias:0', 'bert/encoder/layer_6/attention/self/key/kernel:0', 'bert/encoder/layer_6/attention/self/key/bias:0', 'bert/encoder/layer_6/attention/self/value/kernel:0', 'bert/encoder/layer_6/attention/self/value/bias:0', 'bert/encoder/layer_6/attention/output/dense/kernel:0', 'bert/encoder/layer_6/attention/output/dense/bias:0', 'bert/encoder/layer_6/attention/output/LayerNorm/gamma:0', 'bert/encoder/layer_6/attention/output/LayerNorm/beta:0', 'bert/encoder/layer_6/intermediate/kernel:0', 'bert/encoder/layer_6/intermediate/bias:0', 'bert/encoder/layer_6/output/dense/kernel:0', 'bert/encoder/layer_6/output/dense/bias:0', 'bert/encoder/layer_6/output/LayerNorm/gamma:0', 'bert/encoder/layer_6/output/LayerNorm/beta:0', 'bert/encoder/layer_7/attention/self/query/kernel:0', 'bert/encoder/layer_7/attention/self/query/bias:0', 'bert/encoder/layer_7/attention/self/key/kernel:0', 'bert/encoder/layer_7/attention/self/key/bias:0', 'bert/encoder/layer_7/attention/self/value/kernel:0', 'bert/encoder/layer_7/attention/self/value/bias:0', 'bert/encoder/layer_7/attention/output/dense/kernel:0', 'bert/encoder/layer_7/attention/output/dense/bias:0', 'bert/encoder/layer_7/attention/output/LayerNorm/gamma:0', 'bert/encoder/layer_7/attention/output/LayerNorm/beta:0', 'bert/encoder/layer_7/intermediate/kernel:0', 'bert/encoder/layer_7/intermediate/bias:0', 'bert/encoder/layer_7/output/dense/kernel:0', 'bert/encoder/layer_7/output/dense/bias:0', 'bert/encoder/layer_7/output/LayerNorm/gamma:0', 'bert/encoder/layer_7/output/LayerNorm/beta:0', 'bert/encoder/layer_8/attention/self/query/kernel:0', 'bert/encoder/layer_8/attention/self/query/bias:0', 'bert/encoder/layer_8/attention/self/key/kernel:0', 'bert/encoder/layer_8/attention/self/key/bias:0', 'bert/encoder/layer_8/attention/self/value/kernel:0', 'bert/encoder/layer_8/attention/self/value/bias:0', 'bert/encoder/layer_8/attention/output/dense/kernel:0', 'bert/encoder/layer_8/attention/output/dense/bias:0', 'bert/encoder/layer_8/attention/output/LayerNorm/gamma:0', 'bert/encoder/layer_8/attention/output/LayerNorm/beta:0', 'bert/encoder/layer_8/intermediate/kernel:0', 'bert/encoder/layer_8/intermediate/bias:0', 'bert/encoder/layer_8/output/dense/kernel:0', 'bert/encoder/layer_8/output/dense/bias:0', 'bert/encoder/layer_8/output/LayerNorm/gamma:0', 'bert/encoder/layer_8/output/LayerNorm/beta:0', 'bert/encoder/layer_9/attention/self/query/kernel:0', 'bert/encoder/layer_9/attention/self/query/bias:0', 'bert/encoder/layer_9/attention/self/key/kernel:0', 'bert/encoder/layer_9/attention/self/key/bias:0', 'bert/encoder/layer_9/attention/self/value/kernel:0', 'bert/encoder/layer_9/attention/self/value/bias:0', 'bert/encoder/layer_9/attention/output/dense/kernel:0', 'bert/encoder/layer_9/attention/output/dense/bias:0', 'bert/encoder/layer_9/attention/output/LayerNorm/gamma:0', 'bert/encoder/layer_9/attention/output/LayerNorm/beta:0', 'bert/encoder/layer_9/intermediate/kernel:0', 'bert/encoder/layer_9/intermediate/bias:0', 'bert/encoder/layer_9/output/dense/kernel:0', 'bert/encoder/layer_9/output/dense/bias:0', 'bert/encoder/layer_9/output/LayerNorm/gamma:0', 'bert/encoder/layer_9/output/LayerNorm/beta:0', 'bert/encoder/layer_10/attention/self/query/kernel:0', 'bert/encoder/layer_10/attention/self/query/bias:0', 'bert/encoder/layer_10/attention/self/key/kernel:0', 'bert/encoder/layer_10/attention/self/key/bias:0', 'bert/encoder/layer_10/attention/self/value/kernel:0', 'bert/encoder/layer_10/attention/self/value/bias:0', 'bert/encoder/layer_10/attention/output/dense/kernel:0', 'bert/encoder/layer_10/attention/output/dense/bias:0', 'bert/encoder/layer_10/attention/output/LayerNorm/gamma:0', 'bert/encoder/layer_10/attention/output/LayerNorm/beta:0', 'bert/encoder/layer_10/intermediate/kernel:0', 'bert/encoder/layer_10/intermediate/bias:0', 'bert/encoder/layer_10/output/dense/kernel:0', 'bert/encoder/layer_10/output/dense/bias:0', 'bert/encoder/layer_10/output/LayerNorm/gamma:0', 'bert/encoder/layer_10/output/LayerNorm/beta:0', 'bert/encoder/layer_11/attention/self/query/kernel:0', 'bert/encoder/layer_11/attention/self/query/bias:0', 'bert/encoder/layer_11/attention/self/key/kernel:0', 'bert/encoder/layer_11/attention/self/key/bias:0', 'bert/encoder/layer_11/attention/self/value/kernel:0', 'bert/encoder/layer_11/attention/self/value/bias:0', 'bert/encoder/layer_11/attention/output/dense/kernel:0', 'bert/encoder/layer_11/attention/output/dense/bias:0', 'bert/encoder/layer_11/attention/output/LayerNorm/gamma:0', 'bert/encoder/layer_11/attention/output/LayerNorm/beta:0', 'bert/encoder/layer_11/intermediate/kernel:0', 'bert/encoder/layer_11/intermediate/bias:0', 'bert/encoder/layer_11/output/dense/kernel:0', 'bert/encoder/layer_11/output/dense/bias:0', 'bert/encoder/layer_11/output/LayerNorm/gamma:0', 'bert/encoder/layer_11/output/LayerNorm/beta:0'] when minimizing the loss.`

Can anyone help me, THX!

**Comment by kpe** [comment]
@fakerbrother - can you post the code used to create the model? Is it the same as in #44? If yes, try something like (instead of what's in #44 )
```
self.model_dir ="PretrainedModels/multilingual_L-12_H-768_A-12" 
self.max_seq_len = 64
bert_params = bert.params_from_pretrained_ckpt(self.model_dir)
l_bert = bert.BertModelLayer.from_params(bert_params, name="bert")

self.model = keras.Sequential([
    keras.layers.InputLayer(input_shape=(self.max_seq_len,), dtype='int32'),
    l_bert,
    keras.layers.GlobalAveragePooling1D()
])

self.model.build(input_shape=(None, self.max_seq_len))
bert_ckpt_file   = os.path.join(self.model_dir, "bert_model.ckpt")
bert.load_stock_weights(l_bert, bert_ckpt_file)
```

Please, reopen, if this does not help.

**Comment by fakerbrother** [comment]
Thanks for your response，this problem I have resolved. I used to using PyTorch, I am a rookie in TensorFlow2. So I build model by below code:
import tensorflow.keras as keras

import bert


class BertModel(keras.Model):
    def __init__(self, model_dir, max_seq_len, num_labels=6):
        super(BertModel, self).__init__()
        self.model_dir = model_dir
        self.max_seq_len = max_seq_len
        self.bert_layer = self._get_bert_layer()
        self.flatten_layer = keras.layers.Flatten()
        self.linear = keras.layers.Dense(units=num_labels, input_shape=(768*max_seq_len, ), activation='sigmoid')

    def _get_bert_layer(self):
        bert_params = bert.params_from_pretrained_ckpt(self.model_dir)
        bert_layer = bert.BertModelLayer.from_params(bert_params, name="bert")
        return bert_layer

    def call(self, inputs, training=None, mask=None):
        bert_output = self.bert_layer(inputs)
        flatten_output = self.flatten_layer(bert_output)
        logits = self.linear(flatten_output)
        return logitsmodel = BertModel(bert_path, max_seq_len)
model.build(input_shape=(None, max_seq_len))
bert_ckpt_file = os.path.join(bert_path, "bert_model.ckpt")
bert.load_stock_weights(model.bert_layer, bert_ckpt_file)
Finally, thanks for your help, bset wishes, :) !




------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;"kpe"<notifications@github.com&gt;;
发送时间:&nbsp;2020年1月22日(星期三) 上午8:34
收件人:&nbsp;"kpe/bert-for-tf2"<bert-for-tf2@noreply.github.com&gt;;
抄送:&nbsp;"李小李"<960829614@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [kpe/bert-for-tf2] Gradients do not exist for variables (#46)




@fakerbrother - can you post the code used to create the model? Is it the same as in #44? If yes, try something like (instead of what's in #44 )
 self.model_dir ="PretrainedModels/multilingual_L-12_H-768_A-12"  self.max_seq_len = 64 bert_params = bert.params_from_pretrained_ckpt(self.model_dir) l_bert = bert.BertModelLayer.from_params(bert_params, name="bert") self.model = keras.Sequential([     keras.layers.InputLayer(input_shape=(self.max_seq_len,), dtype='int32'),     l_bert,     keras.layers.GlobalAveragePooling1D() ]) self.model.build(input_shape=(None, self.max_seq_len)) bert_ckpt_file   = os.path.join(self.model_dir, "bert_model.ckpt") bert.load_stock_weights(l_bert, bert_ckpt_file)  
Please, reopen, if this does not help.
